### PR TITLE
ADR about the usage of the TLS registry in client extensions establishing TLS connection

### LINKED ISSUE #41042: Write an ADR explaining how clients should use the TLS registry
### Description

See https://github.com/quarkusio/quarkus/pull/40990#discussion_r1629719466 to see the rationale.

So, we should use a named configuration for the clients. Otherwise, it will likely conflict with servers. Note that multiple clients (like the REST client and the mailer) may use the same TLS configuration - just not the default one.

Follow up from https://github.com/quarkusio/quarkus/pull/40990.

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by cescoffier**
Fix https://github.com/quarkusio/quarkus/issues/41042.

### CODE REVIEW

File: adr/0004-using-the-tls-registry-for-clients.adoc [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+The TLS registry is a way to configure TLS settings in a central place and reuse them across the application and extensions.
+This is particularly useful when an application needs to connect to multiple services using TLS.
+
+Until now, each extension has _maintained_ its own way to configure TLS settings. This leads to a lot of duplication and inconsistencies.
+It's also difficult for the user to understand how to configure TLS settings in Quarkus.
+The TLS registry is a way to solve this problem.
+
+The goal of this ADR is to explain how the registry should be used by client establishing a connection to a server using TLS.
+
+It should be noted that TLS is a sensible and complex topic.
```
Comment (geoand) [inline adr/0004-using-the-tls-registry-for-clients.adoc:19]:
```suggestion
It should be noted that TLS is a sensitive and complex topic.
```

File: adr/0004-using-the-tls-registry-for-clients.adoc [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+The TLS registry is not a silver bullet and does not solve all the problems related to TLS.
+It is a way to simplify the configuration of TLS settings in Quarkus applications.
+
+== The TLS registry
+
+The TLS registry is an extension processing the `quarkus.tls` configuration and proposing a runtime API to access it.
+Note that the configuration and the runtime API are different.
+The configuration is used to define the TLS settings, while the runtime API is used to access these pre-processed settings.
+
+Under `quarkus.tls` can be configured the default configuration, but also _named_ configurations (`quarkus.tls.my-name`).
```
Comment (geoand) [inline adr/0004-using-the-tls-registry-for-clients.adoc:29]:
```suggestion
When using the options directly under `quarkus.tls`, one configures the default (unamed) configuration, while using options under `quarkus.tls.my-name` configures a _named_ configurations.
```

**Comment by cescoffier** [comment]
@geoand As you did the exercise, is there anything I forgot to mention?

**Comment by geoand** [comment]
Will let you know in a few minutes once I've read the whole thing :)

File: adr/0004-using-the-tls-registry-for-clients.adoc [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+At runtime, the https://github.com/quarkusio/quarkus/blob/main/extensions/tls-registry/runtime/src/main/java/io/quarkus/tls/TlsConfigurationRegistry.java[TlsConfigurationRegistry] CDI bean gives access to the default and named _TlsConfiguration_.
+These configurations have been verified during the application startup (password, alias, validity...) and are ready to be used.
+https://github.com/quarkusio/quarkus/blob/main/extensions/tls-registry/runtime/src/main/java/io/quarkus/tls/TlsConfiguration.java[TLSConfiguration] gives access to the key store, trust store, Vert.x options, tailored `SSLContext`...
+It also supports reloading the certificates.
+
+== Configuring clients with the TLS registry
+
+When configuring a _client_ establishing a connection to a server using TLS, the TLS registry should be used to configure the TLS settings.
+The TLS settings must be defined in the `quarkus.tls.<name>` configuration.
```
Comment (geoand) [inline adr/0004-using-the-tls-registry-for-clients.adoc:41]:
```suggestion
The TLS settings must be defined in the `quarkus.tls.<name>.` configuration.
```

File: adr/0004-using-the-tls-registry-for-clients.adoc [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+                + config.tlsConfigurationName.get() + " for the <enter client name> " + name + ".");
+    }
+    configuration = maybeConfiguration.get();
+} else if (tlsRegistry.getDefault().isPresent() && tlsRegistry.getDefault().get().isTrustAll()) {
+    defaultTrustAll = tlsRegistry.getDefault().get().isTrustAll();
+    if (defaultTrustAll) {
+        LOGGER.warn("The default TLS configuration is set to trust all certificates. This is a security risk."
+                + "Please use a named TLS configuration for the mailer " + name + " to avoid this warning.");
+    }
+}
```
Comment (geoand) [inline adr/0004-using-the-tls-registry-for-clients.adoc:125]: We have `TlsConfiguration#from` that does the work of obtaining the proper `TlsConfiguration`

File: adr/0004-using-the-tls-registry-for-clients.adoc [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+----
+
+Also, using the _default_ `trust-all` should be avoided, and a warning should be logged if it's used.
+
+=== Configuring a Vert.x client
+
+When configuring a Vert.x client (HTTP client, SMTP client, gRPC client, etc.), the `TlsConfiguration` should be used to configure the client:
+
+- The `keyStore` and `trustStore` should be set on the `options.setTrustOptions(config.getTrustStoreOptions())` and `options.setKeyCertOptions(config.getKeyStoreOptions())`.
```
Comment (geoand) [inline adr/0004-using-the-tls-registry-for-clients.adoc:134]: I think it's worth mentioning that `keyStore` for clients is used for `mTLS`

File: adr/0004-using-the-tls-registry-for-clients.adoc [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+=== Positive
+
+* A centralized way to configure TLS settings in Quarkus applications.
+* A homogenization of the TLS configuration across Quarkus extensions.
+* A more complete and consistent TLS configuration.
+
+=== Negative
+
+* Dealing with 2 different ways to configure TLS settings in Quarkus applications until the extension-specific configurations are removed.two
```
Comment (geoand) [inline adr/0004-using-the-tls-registry-for-clients.adoc:189]:
```suggestion
* Dealing with 2 different ways to configure TLS settings in Quarkus applications until the extension-specific configurations are removed.
```

**Comment by geoand** [comment]
> Will let you know in a few minutes once I've read the whole thing :)

Read it all, nothing missing that I can think of!

File: adr/0004-using-the-tls-registry-for-clients.adoc (hunk 4 above)
Comment (cescoffier) [reply adr/0004-using-the-tls-registry-for-clients.adoc]: Fixed.

**Review by geoand — APPROVED** [review]

**Pull request merged**